nntrainer/models/neuralnet.cpp

```diff
@@ -112,8 +123,14 @@ int NeuralNetwork::compile() {
     rep.push_back(*iter);
   }
 
-  FlattenRealizer fr;
-  rep = fr.realize(rep);
+  std::vector<std::unique_ptr<GraphRealizer>> realizers;
+
+  realizers.emplace_back(new PreviousInputRealizer(input_layers));
+  realizers.emplace_back(new FlattenRealizer());
+
+  for (auto &realizer : realizers) {
+    rep = realizer->realize(rep);
```

Comment:
consider moving this inside the NetworkGraph class.
As this is modifying (expand/shrink) the graph depending on the properties of its node, its better suited for NetworkGraph class.

Reply:
realizer is happening before being added to network graph. (thus named to GraphRepresentation or you might wanna call this module, layer group etc..). This is because in this way, we can make some stricter assumption of some states of layer node and control the temporal coupling between topological sort and realization. For example, we can guarantee that output connection is not set so we don't have to update output connection.

Ideally, model.addLayer() or such will create GraphRepresentation, not graph at the beginning and model.compile constructs network graph and constructor will sort at the time.

fyi) There already exist some troublesome behavior making it hard to migrate to tensor connection from some realization + loss realization because they are happening after output connections are made.

Reply:
Then in a layer PR, consider making a new class to move this out of neuralnet class.

Reply:
Yeah, but this will have a lot of work required, so unfortunately I have to delay this but added as a todo previously.

Reply:
> Then in a layer PR, consider making a new class to move this out of neuralnet class.

layer PR -> later* PR :)
Sure, not needed now. The comment is just to put to note here.

Reply:
I think it could be one of compilation work. So I think one of the module of compiler might be better.